### ISSUE TITLE
Sidekiq swarm not shutting down gracefully on Heroku

### ISSUE THREAD

**Issue by andrepiske** [sidekiq/sidekiq#5956]
Ruby version: 3.2.2
Rails version: non-rails or 7.0.5
Sidekiq / Pro / Enterprise version(s): 6.5.9/5.5.8/2.5.3 or 7.1.2/7.1.3/7.1.1

We're seeing an issue when using sidekiqswarm on Heroku. We see this in our rails app, which is using sidekiq 6.5.9 (with pro & ent versions like above), but I was able to boil the issue down to a minimal app, without rails, to demonstrate the issue. I published the code at https://github.com/andrepiske/sidekiq-swarm-issue so you can easily reproduce it if you want.

## The issue

Sidekiq won't gracefully shutdown when running via `sidekiqswarm`, but works just fine when running via `sidekiq`. I could only observe this in Heroku only (using private spaces). I'm using AWS Elasticache, redis 6.2.6, as the backing redis store. Networking is: an AWS VPC and Heroku private spaces with peering.
I wasn't able to reproduce this on my macOS nor in docker containers, just in Heroku -- but I believe a considerable amount of folks are using heroku, which would make this issue relevant to many.

My test app (github link above) is as follows: It uses sidekiq-ent's scheduler to run a job every minute: `mgr.register("* * * * *", 'Worker::TpRandomizer', scheduled: true)` ([link to code](https://github.com/andrepiske/sidekiq-swarm-issue/blob/master/config/bootloader.rb#L87)). That class essentially performs the following ([link to code](https://github.com/andrepiske/sidekiq-swarm-issue/blob/master/app/worker/tp_randomizer.rb#L7-L19)):

1. Generates a random number and prints it like in the first line of the log below (`New number: 105052`)
2. Actively waits ([link to code](https://github.com/andrepiske/sidekiq-swarm-issue/blob/master/app/worker/tp_randomizer.rb#L25-L32)) for `ENV['TP_WAIT_FOR']` seconds (I tested with 14 and 21)
3. After the wait, puts that random number in a certain redis key. That's it.

There's an endpoint `/rnd` where you can retrieve that number from the redis key ([link to code](https://github.com/andrepiske/sidekiq-swarm-issue/blob/master/app/tp_app.rb#L4-L15)), so to check if the value actually changed.

So the sidekiq job essentially generates a random number, waits 21 seconds, then puts it into a redis key. The problem occurs when there is a dyno restart (e.g. during a deploy). I'm using `$ heroku config:set -a my-app-name DUMMY=<some-random-value>` to force restarts on Heroku and the issue happens every single time.

In the log below, you can see the job started at second `:00` (note that Sidekiq log level is `debug`). Then just before second `:07`, I force a restart using the heroku command above. At the `:11` mark, the two sidekiq processes (running swarm with 2 processes) get the TERM signal. Note that Heroku sends TERM signals to all processes in the dyno ([docs](https://devcenter.heroku.com/articles/dynos#shutdown)). You can see that at second `:12` everything is already shutdown, but the task takes 21 seconds to run and my configured grace time is of 25 seconds.

```
Jun 29 10:51:00 app[skiq] info New number: 105052
Jun 29 10:51:00 app[skiq] info Waiting 21 before setting number
Jun 29 10:51:00 app[skiq] INFO I, [2023-06-29T13:51:00.008376 #14]  INFO -- : Enqueued periodic job Worker::TpRandomizer with JID 8511564b2ac02e44f53b8165 for 2023-06-29 13:51:00 +0000
Jun 29 10:51:00 app[skiq] DEBUG D, [2023-06-29T13:51:00.008452 #14] DEBUG -- : Tick: [2023-06-29 13:51:00.000140495 +0000, "8e5181d02d9949b803d53aef2c379b2b35ef815a"]
Jun 29 10:51:07 heroku[skiq] info State changed from up to stopping
Jun 29 10:51:11 app[skiq] DEBUG D, [2023-06-29T13:51:11.787576 #14] DEBUG -- : Got TERM signal
Jun 29 10:51:11 app[skiq] DEBUG D, [2023-06-29T13:51:11.787984 #15] DEBUG -- : Got TERM signal
Jun 29 10:51:11 app[skiq] INFO I, [2023-06-29T13:51:11.788132 #15]  INFO -- : Shutting down
Jun 29 10:51:11 app[skiq] INFO I, [2023-06-29T13:51:11.788182 #14]  INFO -- : Shutting down
Jun 29 10:51:11 app[skiq] INFO I, [2023-06-29T13:51:11.788225 #15]  INFO -- : Terminating quiet threads for default capsule
Jun 29 10:51:11 app[skiq] INFO I, [2023-06-29T13:51:11.788404 #14]  INFO -- : Terminating quiet threads for default capsule
Jun 29 10:51:11 app[skiq] INFO I, [2023-06-29T13:51:11.788513 #14]  INFO -- : Scheduler exiting...
Jun 29 10:51:11 app[skiq] INFO I, [2023-06-29T13:51:11.788470 #15]  INFO -- : Scheduler exiting...
Jun 29 10:51:11 app[skiq] DEBUG D, [2023-06-29T13:51:11.788656 #15] DEBUG -- : Firing quiet event
Jun 29 10:51:11 app[skiq] DEBUG D, [2023-06-29T13:51:11.788803 #15] DEBUG -- : Firing shutdown event
Jun 29 10:51:11 app[skiq] DEBUG D, [2023-06-29T13:51:11.789611 #14] DEBUG -- : Firing quiet event
Jun 29 10:51:11 app[skiq] DEBUG D, [2023-06-29T13:51:11.789731 #14] DEBUG -- : Firing shutdown event
Jun 29 10:51:11 app[skiq] INFO I, [2023-06-29T13:51:11.789927 #14]  INFO -- : Periodic subsystem stopped
Jun 29 10:51:12 heroku[skiq] info Process exited with status 0
Jun 29 10:51:12 heroku[skiq] info State changed from stopping to down
```

A note on SuperFetch: it does pick up the job again later, so SuperFetch is working. Which evidences that the job indeed died prematurely.

## Expected behavior

Sidekiq would wait gracefully for 25 seconds ([the configured timeout](https://github.com/andrepiske/sidekiq-swarm-issue/blob/master/config/sidekiq.yml#L4)) before SIGKILLing the job.

## Actual behavior

Sidekiq exits before, even with status 0. But it doesn't seem to really have been a clean shutdown because it's missing many log lines that would evidence that.

**Comment by mperham** [comment]
Thanks for the bug report and reproduction. I'm on sabbatical until August so I won't have time to dive deep into this for a bit. If you determine a root cause and/or a fix, I'd be happy to review and merge.

**Comment by mperham** [comment]
In this cases, it helps to remove as much complexity as possible. Can you still repro the problem if you remove Docker and containers from the scenario?

**Comment by mperham** [comment]
Here's what I see:

```
2023-09-01T18:29:00.000349+00:00 app[skiq.1]: D, [2023-09-01T18:29:00.000267 #5] DEBUG -- : Now: 2023-09-01 18:29:00 +0000, Next: 2023-09-01 18:30:00 +0000
2023-09-01T18:29:00.003737+00:00 app[skiq.1]: I, [2023-09-01T18:29:00.003679 #4]  INFO -- : start
2023-09-01T18:29:00.004102+00:00 app[skiq.1]: New number: 108930
2023-09-01T18:29:00.004132+00:00 app[skiq.1]: Waiting 20 before setting number
2023-09-01T18:29:00.004692+00:00 app[skiq.1]: I, [2023-09-01T18:29:00.004658 #5]  INFO -- : Enqueued periodic job Worker::TpRandomizer with JID 8412892faca5cfb51f01f011 for 2023-09-01 18:29:00 +0000
2023-09-01T18:29:00.004732+00:00 app[skiq.1]: D, [2023-09-01T18:29:00.004716 #5] DEBUG -- : Tick: [2023-09-01 18:29:00.000161879 +0000, "8e5181d02d9949b803d53aef2c379b2b35ef815a"]
2023-09-01T18:29:09.268312+00:00 heroku[skiq.1]: State changed from up to down
2023-09-01T18:29:10.050522+00:00 heroku[skiq.1]: Stopping all processes with SIGTERM
2023-09-01T18:29:10.076577+00:00 app[skiq.1]: D, [2023-09-01T18:29:10.076492 #5] DEBUG -- : Got TERM signal
2023-09-01T18:29:10.076638+00:00 app[skiq.1]: I, [2023-09-01T18:29:10.076616 #5]  INFO -- : Shutting down
2023-09-01T18:29:10.076697+00:00 app[skiq.1]: I, [2023-09-01T18:29:10.076678 #5]  INFO -- : Terminating quiet threads for default capsule
2023-09-01T18:29:10.076775+00:00 app[skiq.1]: I, [2023-09-01T18:29:10.076752 #5]  INFO -- : Scheduler exiting...
2023-09-01T18:29:10.076839+00:00 app[skiq.1]: D, [2023-09-01T18:29:10.076819 #5] DEBUG -- : Firing quiet event
2023-09-01T18:29:10.077400+00:00 app[skiq.1]: D, [2023-09-01T18:29:10.077371 #5] DEBUG -- : Firing shutdown event
2023-09-01T18:29:10.077943+00:00 app[skiq.1]: D, [2023-09-01T18:29:10.077871 #4] DEBUG -- : Got TERM signal
2023-09-01T18:29:10.078048+00:00 app[skiq.1]: I, [2023-09-01T18:29:10.078020 #4]  INFO -- : Shutting down
2023-09-01T18:29:10.078123+00:00 app[skiq.1]: I, [2023-09-01T18:29:10.078098 #4]  INFO -- : Terminating quiet threads for default capsule
2023-09-01T18:29:10.078335+00:00 app[skiq.1]: I, [2023-09-01T18:29:10.078294 #4]  INFO -- : Scheduler exiting...
2023-09-01T18:29:10.078355+00:00 app[skiq.1]: I, [2023-09-01T18:29:10.078315 #5]  INFO -- : Leader stepping down
2023-09-01T18:29:10.078563+00:00 app[skiq.1]: D, [2023-09-01T18:29:10.078526 #4] DEBUG -- : Firing quiet event
2023-09-01T18:29:10.078797+00:00 app[skiq.1]: I, [2023-09-01T18:29:10.078599 #5]  INFO -- : Periodic subsystem stopped
2023-09-01T18:29:10.078798+00:00 app[skiq.1]: D, [2023-09-01T18:29:10.078626 #4] DEBUG -- : Firing shutdown event
2023-09-01T18:29:10.579076+00:00 app[skiq.1]: I, [2023-09-01T18:29:10.578964 #4]  INFO -- : Pausing to allow jobs to finish...
2023-09-01T18:29:10.579174+00:00 app[skiq.1]: I, [2023-09-01T18:29:10.579127 #5]  INFO -- : Pausing to allow jobs to finish...
2023-09-01T18:29:11.082080+00:00 app[skiq.1]: D, [2023-09-01T18:29:11.082002 #5] DEBUG -- : SuperFetch[default]: Unregistered super queues ["queue:sq|skiq.1:5:c4b39c507fa1|default"]
2023-09-01T18:29:11.083330+00:00 app[skiq.1]: I, [2023-09-01T18:29:11.083299 #5]  INFO -- : Bye!
2023-09-01T18:29:19.821758+00:00 app[skiq.1]: Done waiting!
2023-09-01T18:29:19.826673+00:00 app[skiq.1]: I, [2023-09-01T18:29:19.826606 #4]  INFO -- : done
2023-09-01T18:29:20.083839+00:00 app[skiq.1]: D, [2023-09-01T18:29:20.083769 #4] DEBUG -- : SuperFetch[default]: Unregistered super queues ["queue:sq|skiq.1:4:a2a395d6067a|default"]
2023-09-01T18:29:20.086440+00:00 app[skiq.1]: I, [2023-09-01T18:29:20.086400 #4]  INFO -- : Bye!
2023-09-01T18:29:20.243324+00:00 heroku[skiq.1]: Process exited with status 0
```

That looks basically correct to me. There's two child workers, 4 and 5. 5 is quiet and exits almost immediately. 4 waits, finishes its job and exits almost immediately thereafter.

**Comment by mperham** [comment]
I'll also note that there has been a long-term Ruby bug just discovered affecting `Process.waitpid` and child processes which can cause process hangs. It's fixed in Ruby 3.3 but I'm not clear which versions of 3.2 or 3.1 have a fix backported. 

If you see issues with the stability of sidekiqswarm, I would test Ruby 3.3.

https://github.com/puma/puma/pull/3314
